addon-test-support/-private/mock/proxy.js

```diff
@@ -65,8 +57,19 @@ export default function proxyFactory(original, makeHolder = () => ({})) {
       return true;
     },
     defineProperty(target, key, descriptor) {
-      assertConfigurableDescriptor(target, key);
+      if (!(key in config.holder)) {
+        // First copy the original descriptor to a dummy object.
+        const originalDescriptor = Object.getOwnPropertyDescriptor(target, key);
+        const dummy = {};
+        Object.defineProperty(dummy, key, originalDescriptor);
+
+        // Then try overriding it with the new descriptor, which might throw a TypeError.
+        Object.defineProperty(config.holder, key, descriptor);
```

Comment:
This is exactly the same code as 4 lines below, so what purpose does that have? And how does that help when it would throw a TypeError, as we don't catch anything?

Reply:
Right! It should be `dummy` instead of `config.holder`.

What this is supposed to do is check whether the original descriptor of the `target` is actually overridable. If it is not overridable, this line will throw an error. I did not explicitly catch that error as throwing an error in or returning `false` from `Proxy#defineProperty` will lead to the same result basically. However, to make it more explicit and to not rely on behavior which is not explicitly defined, I'll wrap this in a `try / catch` and return `false`.

The reason that we do this in the first place is that `defineProperty` will be executed, even if one of the [invariant scenarios](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/handler/defineProperty#Invariants) is hit. It _has_ to return `false` though. If it does not, a `TypeError` will be raised.

L71 `Object.defineProperty(config.holder, key, descriptor);` would have been executed nevertheless though, which means that the `defineProperty` did succeed anyway. I will add some tests around this.

Reply:
Hm, while writing the tests, I noticed some more inconsistencies. I guess I'll have to manually codify all invariants and some more rules. 🙈